text/0000-forget-marker-trait.md

```diff
@@ -0,0 +1,803 @@
+
+Earlier it was stated that Bounds for `Self` and associated types should default to `?Forget` instead of `Forget`. This is due to an important case. If a user of the library updated earlier than the library, then without that change it will observe that associated types of traits are `Forget`, so it would be a breaking change for the library to lift that constraint in the future. But now, the user will observe `?Forget`, thus it cannot rely on them being `Forget`. But for users that did not migrate, as well as the library itself, it will not be observable due to `default_generic_bounds` still being `Forget`.
+
+```rust
+// This indicates that user explicity opted in
+#![default_generic_bounds(?Forget)]
+
+// After opting in, user needs to add `T::baz(..): Forget` to silence the error - quite easy.
+async fn foo<T: other_crate::Trait>(bar: T) {
+    let fut = bar.baz();
+    // Compiler will emit an error, as `fut` maybe `!Forget`, because we set `default_generic_bounds`
+    // to `?Forget`, and default for associated types in `other_crate` is already `?Forget`. Otherwise it
+    // would have been a breaking change for `other_crate` to make future provided by `baz` `!Forget`,
+    // as this code would've compiled now but not in the future.
+    core::mem::forget(fut);
+}
+
+// A library that did not migrate yet. `Trait::bar(..)` is not locked into `Forget`, but
+// this `other_crate` and other crates can only observe `Trait::bar(..): Forget` cases.
+mod other_crate {
+    trait Trait {
+        async fn baz();
+    }
+}
+```
+
+#### Macros
+
+If macro-library generates code, some problems during the migration are possible:
+
+```rust
+mod user {
+    #![default_generic_bounds(?Forget)]
+
+    ::library::make!(); // Will not compile because `T` is `?Forget`.
+}
+
+mod user {
+    #[macro_export]
+    macro_rules! make {
+        () => {
+            pub fn foo<T>(t: T) {
+                ::core::mem::forget(t);
+            }
+        }
+    }
+}
+```
+
+#### Changing default
+
+It is not required, but in next editions we may swap the default for `default_generic_bounds`. Crates that want to continue using old default in next editions will set `#![default_generic_bounds(Forget)]`.
+
+### Migration over the edition with [default auto traits]
+[edition-migration-with-mask]: #edition-migration-with-mask
+
+If [`local_default_bounds`] is not accepted, we can make a satisfactory migration by having editions <= 2024 have `Forget` as the default, and editions after 2024 have `?Forget` as the default.
+
+While it will not split the ecosystem, it will require everyone to make a migration just as in the [`local_default_bounds`] solution. There are concern about locking crates into `Forget` bounds on associated types in traits (like `async fn`). Migration can probably be automated for `#![forbid(unsafe)]` crates.
```

Comment:
It's the absence of unsafe code that's the determining factor, not a specific lint against it, right?
(The migration tool could run the code for that lint before migrating itself, for example.)

```suggestion
While it will not split the ecosystem, it will require everyone to make a migration just as in the [`local_default_bounds`] solution. There are concern about locking crates into `Forget` bounds on associated types in traits (like `async fn`). Migration can probably be automated for crates without unsafe code.
```